### PR TITLE
refactor: move python server to lang package as runner component

### LINKED ISSUE #839: Runner as service in language component
In order to implement #37, the language component needs to have access to the runner to infer the initial schema of a dataset that gets loaded from a file. Hence, the runner access has to be moved from the VS Code component to the language component. 

There, it can be added as another service. The path to the runner can be passed when creating a service module as an optional argument. The system should still work if no path is passed (minus the runner of course).

### PR DESCRIPTION
**Pull request by WinPlay02**
Closes #839

- refactored python server into runner component

### CODE REVIEW

File: packages/safe-ds-lang/src/language/safe-ds-module.ts [hunk 1]
```diff
@@ -150,6 +154,9 @@ export const SafeDsModule: Module<SafeDsServices, PartialLangiumServices & SafeD
         PackageManager: (services) => new SafeDsPackageManager(services),
         SettingsProvider: (services) => new SafeDsSettingsProvider(services),
     },
+    runtime: {
```
Comment (WinPlay02) [inline packages/safe-ds-lang/src/language/safe-ds-module.ts:157]: I suppose the optional parameter should be passed somewhere here?
https://github.com/Safe-DS/DSL/blob/21298d0a3be7c9e0994614641471cb0344826696/packages/safe-ds-lang/src/language/safe-ds-module.ts#L209

**Comment by WinPlay02** [comment]
@lars-reimann Almost all the functions require a runner process, which would complicate the testing setup. How should I proceed there?

**Comment by lars-reimann** [comment]
> @lars-reimann Almost all the functions require a runner process, which would complicate the testing setup. How should I proceed there?

Which functions do you mean?

File: packages/safe-ds-lang/src/language/runner/safe-ds-runner.ts [hunk 2]
```diff
@@ -177,32 +202,39 @@ export class SafeDsRunner {
...
+    /* c8 ignore stop */
 
     /**
      * Remove information from all previous pipeline executions.
      */
+    /* c8 ignore start */
     public dropAllPipelineExecutionContexts() {
         this.executionInformation.clear();
     }
+    /* c8 ignore stop */
```
Comment (lars-reimann) [inline packages/safe-ds-lang/src/language/runner/safe-ds-runner.ts]: You can also remove the intermediate stop/start commands and keep only the beginning and end ones. I wouldn't mind either, if you just put a `/* c8 ignore start */` at the start of the file with an explanatory comment.

File: packages/safe-ds-lang/tests/language/runner/messages.test.ts
Comment (WinPlay02) [inline packages/safe-ds-lang/tests/language/runner/messages.test.ts:1]: These tests should cover everything in messages.ts, but this test file does not seem to get executed on CI?

Comment (WinPlay02) [reply packages/safe-ds-lang/tests/language/runner/messages.test.ts:1]: I think the wrong coverage was caused by an incorrect import statement

File: packages/safe-ds-lang/src/language/safe-ds-module.ts [hunk 3]
```diff
@@ -204,13 +216,17 @@ export const createSafeDsServices = function (context: DefaultSharedModuleContex
  *  - Services specified in this file
  *
  * @param context Optional module context with the LSP connection.
+ * @param runnerCommand Optional command to start the runner, if available.
  * @return An object wrapping the shared services and the language-specific services.
  */
-export const createSafeDsServicesWithBuiltins = async function (context: DefaultSharedModuleContext): Promise<{
+export const createSafeDsServicesWithBuiltins = async function (
+    context: DefaultSharedModuleContext,
+    runnerCommand: string | undefined = undefined,
```
Comment (lars-reimann) [inline packages/safe-ds-lang/src/language/safe-ds-module.ts:224]: For the sake of future extensibility, I'd suggest wrapping the `runnerCommand` into an options object:

```ts
export const createSafeDsServicesWithBuiltins = async function (
    context: DefaultSharedModuleContext,
    options?: ModuleOptions
) // ...

interface ModuleOptions {
    runnerCommand?: string
}
```

File: packages/safe-ds-lang/src/language/safe-ds-module.ts [hunk 4]
```diff
@@ -178,16 +185,21 @@ export const SafeDsSharedModule: Module<SafeDsSharedServices, DeepPartial<SafeDs
  *  - Services specified in this file
  *
  * @param context Optional module context with the LSP connection.
+ * @param runnerCommand Optional command to start the runner, if available.
  * @return An object wrapping the shared services and the language-specific services.
  */
-export const createSafeDsServices = function (context: DefaultSharedModuleContext): {
+export const createSafeDsServices = function (
+    context: DefaultSharedModuleContext,
+    runnerCommand: string | undefined = undefined,
```
Comment (lars-reimann) [inline packages/safe-ds-lang/src/language/safe-ds-module.ts:193]: Options object (see `createSafeDsServicesWithBuiltins`).

File: packages/safe-ds-lang/tests/language/runner/safe-ds-runner.test.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        });
+        const foundPort = await runner.findFirstFreePort(portNumber);
+        server.close();
+        expect(foundPort).toStrictEqual(portNumber + 1);
+    });
+    it('findFirstFreePort-available', async () => {
+        const portNumber = 46825;
+        const foundPort = await runner.findFirstFreePort(portNumber);
+        expect(foundPort).toStrictEqual(portNumber);
+    });
```
Comment (lars-reimann) [inline packages/safe-ds-lang/tests/language/runner/safe-ds-runner.test.ts:30]: Wrap the test cases (`it`) in test suites (`describe`) for the individual functions inside the top-level `describe(`SafeDsRunner`)`.

File: packages/safe-ds-lang/tests/language/runner/safe-ds-runner.test.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    it('getMainModuleName-sdstest', async () => {
+        const document = services.shared.workspace.LangiumDocumentFactory.fromString('', URI.file('/a-b c.sdstest'));
+        const mainModuleName = runner.getMainModuleName(document);
+        expect(mainModuleName).toStrictEqual('a_b_c');
+    });
+    it('getMainModuleName-other', async () => {
+        const document = services.shared.workspace.LangiumDocumentFactory.fromString('', URI.file('/a-b c.sdstest2'));
+        const mainModuleName = runner.getMainModuleName(document);
+        expect(mainModuleName).toStrictEqual('a_b_c_sdstest2');
+    });
```
Comment (lars-reimann) [inline packages/safe-ds-lang/tests/language/runner/safe-ds-runner.test.ts:45]: Another block.

File: packages/safe-ds-lang/tests/language/runner/safe-ds-runner.test.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    it('generateCodeForRunner', async () => {
+        const document = services.shared.workspace.LangiumDocumentFactory.fromString(
+            'package a\n\npipeline mainpipeline {}',
+            URI.file('/b.sdstest'),
+        );
+        const [programCodeMap] = runner.generateCodeForRunner(document, undefined);
+        expect(JSON.stringify(programCodeMap).replaceAll('\\r\\n', '\\n')).toStrictEqual(
+            '{"a":{"gen_b":"# Pipelines --------------------------------------------------------------------\\n\\ndef mainpipeline():\\n    pass\\n","gen_b_mainpipeline":"from .gen_b import mainpipeline\\n\\nif __name__ == \'__main__\':\\n    mainpipeline()\\n"}}',
+        );
+    });
```
Comment (lars-reimann) [inline packages/safe-ds-lang/tests/language/runner/safe-ds-runner.test.ts:55]: Another block (so everything is on the same level).

**Review by lars-reimann — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**